src/mlpack/methods/ann/augmented/tasks/copy_impl.hpp

```diff
@@ -0,0 +1,107 @@
+/**
+ * @file copy_impl.hpp
+ * @author Konstantin Sidorov
+ *
+ * Implementation of CopyTask class
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_AUGMENTED_TASKS_COPY_IMPL_HPP
+#define MLPACK_METHODS_AUGMENTED_TASKS_COPY_IMPL_HPP
+
+// In case it hasn't been included yet.
+#include "copy.hpp"
+
+namespace mlpack {
+namespace ann /* Artificial Neural Network */ {
+namespace augmented /* Augmented neural network */ {
+namespace tasks /* Task utilities for augmented */ {
+
+CopyTask::CopyTask(const size_t maxLength,
+                   const size_t nRepeats,
+                   const bool addSeparator) :
+    maxLength(maxLength),
+    nRepeats(nRepeats),
+    addSeparator(addSeparator)
+{
+  if (maxLength <= 1) 
+  {
+    std::ostringstream oss;
+    oss << "CopyTask::CopyTask(): maximum sequence length ("
+        << maxLength << ") "
+        << "should be at least 2!"
+        << std::endl;
+    throw std::invalid_argument(oss.str());
+  }
+  if (nRepeats <= 0)
+  {
+    std::ostringstream oss;
+    oss << "CopyTask::CopyTask(): repetition count (" << nRepeats << ") "
+        << "is not positive!"
+        << std::endl;
+    throw std::invalid_argument(oss.str());
+  }
+  // Just storing task-specific parameters.
+}
+
+const void CopyTask::Generate(arma::field<arma::mat>& input,
+                              arma::field<arma::mat>& labels,
+                              const size_t batchSize,
+                              bool fixedLength)
+{
+  input = arma::field<arma::mat>(batchSize);
+  labels = arma::field<arma::mat>(batchSize);
+  size_t size = maxLength;
+  for (size_t i = 0; i < batchSize; ++i)
+  {
+    if (!fixedLength)
+    {
+      // Generate random uniform length from [2..maxLength].
+      size = mlpack::math::RandInt(2, maxLength+1);
+    }
+    arma::colvec vecInput = arma::randi<arma::colvec>(
+      size, arma::distr_param(0, 1));
+    arma::colvec vecLabel = arma::conv_to<arma::colvec>::from(
+      arma::repmat(vecInput, nRepeats, 1));
+    size_t totSize = vecInput.n_elem + addSeparator + vecLabel.n_elem;
+    input(i) = arma::zeros(totSize, 2);
+    input(i).col(0).rows(0, vecInput.n_elem-1) =
+      vecInput;
+    if (addSeparator)
+      input(i).at(vecInput.n_elem, 0) = 0.5;
```

Comment:
Nope, everything is OK here - I checked it by throwing some `std::cerr` output. As @zoq pointed out earlier, we have column-major representation here, so we store sequences as `N * 2` matrix, where `N` is the sequence length. (We use 2 columns to store the sequence element AND information on whether this cell should be used for writing model's output)

For separating input part and output part, we use special value not used anywhere else - `0.5`. (We discussed with @zoq whether should we set 1 in the second slot of the separator and decided not to, because the separator is not used for storing the original sequence)